contracts/v2/core/WithdrawalQueue.sol

```diff
@@ -0,0 +1,165 @@
+    mapping(address => uint256) public redeemRequests;
+
+    event RedeemRequest(
+        address indexed requester,
+        address indexed owner,
+        uint256 indexed requestId,
+        address operator,
+        uint256 assets
+    );
+    event Redeem(address indexed requester, address indexed receiver, uint256 shares, uint256 assets);
+
+    constructor(address _minter, address _wsgEth, uint256 _epochLength) FIFOQueue(_epochLength) {
+        MINTER = _minter;
+        WSGETH = _wsgEth;
+
+        uint256 maxUint256 = 2 ** 256 - 1;
+
+        IERC20(WSGETH).approve(_minter, maxUint256);
+
+        _grantRole(GOV, msg.sender);
+    }
+
+    function requestRedeem(
+        uint256 shares,
+        address requester,
+        address owner
+    ) external onlyOwnerOrOperator(owner) nonReentrant whenNotPaused returns (uint256 requestId) {
+        if (shares == 0) {
+            revert Errors.InvalidAmount();
+        }
+        IERC20(WSGETH).transferFrom(owner, address(this), shares); // asset here is the Vault underlying asset
+
+        requestId = requestsPending++;
+        requests[requestId] = Request({requester: requester, shares: shares});
+
+        _stakeForWithdrawal(owner, IERC4626(WSGETH).previewRedeem(shares));
+        totalPendingRequest += shares;
+        redeemRequests[requester] += shares; // underflow would revert if not enough claimable shares
+
+        emit RedeemRequest(requester, owner, requestId, msg.sender, shares);
+        return requestId;
+    }
+
+    function redeem(
+        uint256 shares,
+        address receiver,
+        address requester
+    ) external onlyOwnerOrOperator(requester) nonReentrant whenNotPaused returns (uint256 assets) {
+        if (shares == 0) {
+            revert Errors.InvalidAmount();
+        }
+
+        assets = IERC4626(WSGETH).previewRedeem(shares);
+
+        if (claimableRedeemRequest(requester) < assets) {
+            revert Errors.InvalidAmount();
+        }
+        _withdraw(requester, assets);
+
+        // Treat everything as claimableRedeemRequest and validate here if there's adequate funds
+        redeemRequests[requester] -= shares; // underflow would revert if not enough claimable shares
+        totalPendingRequest -= shares;
+
+        uint256 queueBalance = address(this).balance;
+        uint256 minterBalance = MINTER.balance;
+
+        if (queueBalance + minterBalance < assets) {
+            revert Errors.InsufficientBalance();
+        }
+
+        // Track total returned
+        totalAssetsOut += assets;
+        requestsFulfilled++;
+        // This feels suboptimal, but is the easiest way to always burn the token on redemptions
+        if (assets > minterBalance) {
+            uint256 diff = assets - minterBalance;
+            // We need to use donate/transfer etc. cant deposit and mint more shares as that messes up accouting
+            SharedDepositMinterV2(payable(MINTER)).deposit{value: diff}();
```

Comment:
@devlancer412 can you please see why this doesnt work? 
looks to only work with deposit call, but we want to call 
`
            SharedDepositMinterV2(payable(MINTER)).donate{value: diff}();
`
or `Address.sendValue` etc, anything that allows us to increase the holdings in the target minter without minting any new shares to this withdrawalQueue contract, as the users minted shares should be held in this contract, then burnt from it when the call is done.  
Otherwise the accounting gets wonky.

Reply:
Did you fix this?

Reply:
I think it will create a dummy sgEth on the withdrawQ contract. So I think it would be better to just transfer for now.

Reply:
But I'm not sure where is 32 eth chunk logic. is it implemented?